Expose wgs84ToWindowCoordinates on the Scene

bampakoa opened the pull request:
Expose SceneTransforms.wgs84ToWindowCoordinates in the Scene prototype. Fixes #2645

@mramato Please notice that I have not remove it from `SceneTransforms`, as I did not know if this is the desired behavior.

bampakoa commented:
This is ready.

hpinkos commented:
Thanks @bampakoa!  This looks good to me, @mramato can you take a quick look and merge if this is what you had in mind?

hpinkos commented:
Actually, one comment.  Your thumbnail for the sandcastle example is a little large.  Use the 'View as thumbnail' button at the top of Sandcastle to make the canvas the correct size for the thumbnail

mramato commented on Source/Scene/Scene.js:
```diff
@@ -3311,5 +3311,27 @@ define([
+     *     console.log(scene.wgs84ToWindow(position));
+     * }, Cesium.ScreenSpaceEventType.MOUSE_MOVE);
+     */
+    Scene.prototype.wgs84ToWindow = function(position, result) {
```
My initial suggestion was going to be to keep `wgs84ToWindowCoordinates` as the name, but looking into it some more, I think having `wgs84` in the name at all is actually incorrect.  I don't believe this always uses WGS84 and instead uses the Scene's ellipsoid (which may be something non-wgs84).  @bagnell can you confirm this?

If my assumption is correct, something like `scene.computeCanvasCoordinates` may be a better choice (and obviously we would need to update the documentation to match.  Something like "The position in world coordinates" or the 
"position in the scene's ellipsoid".

Anyone else have any thoughts here?

bagnell replied:
Yes, it uses the Scene's ellipsoid. I believe at one time it was always using the WGS84 ellipsoid but we never changed the name.

emackey replied:
How about `ellipsoidToWindowCoordinates` ?

mramato replied:
@emackey I think that name is confusing because the coordinates don't have to be on the ellipsoid, this function works with any position.  My only other thought was something like `fixedFrameToCanvasCoordinates` but I think it's a bit obtuse.  I like `computeCanvasCoordinates` because I think it's more discoverable.  Anyone else have any ideas?

emackey replied:
`cartesianToWindowCoordinates` ?

mramato replied:
I think I like `cartesianToCanvasCoordinates` since the coordinates are relative to the canvas.  Window implies the entire window object. (plus, alliteration is always awesome).

mramato commented on Specs/Scene/SceneSpec.js:
```diff
@@ -1122,4 +1122,13 @@ defineSuite([
     });
 
+    it('converts to window coordinates',function(){
+        spyOn(SceneTransforms, 'wgs84ToWindowCoordinates').and.callThrough();
```
No need to have `and.callThrough` here.

mramato commented on Specs/Scene/SceneSpec.js:
```diff
@@ -1122,4 +1122,13 @@ defineSuite([
+    it('converts to window coordinates',function(){
+        spyOn(SceneTransforms, 'wgs84ToWindowCoordinates').and.callThrough();
+
+        expect(scene).toRenderAndCall(function() {
```
You can then remove this line because the test can be a lot simpler.

mramato commented on Specs/Scene/SceneSpec.js:
```diff
@@ -1122,4 +1122,13 @@ defineSuite([
+
+        expect(scene).toRenderAndCall(function() {
+            scene.wgs84ToWindow(new Cartesian3());
+            expect(SceneTransforms.wgs84ToWindowCoordinates).toHaveBeenCalled();
```
Use `toHaveBeenCalledWith` so that you can pass the expected arguments (and test using a result parameter).

bampakoa commented:
@mramato I have finished with your recommended changes in the test file, so I am waiting for you guys for the name of the function.

emackey commented:
`cartesianToCanvasCoordinates` it is.

bampakoa commented:
@mramato @emackey I have changed the name of the method along with the tests, the name of the SandCastle demo and I have updated the documentation accordingly.

pjcozzi commented:
Thanks @bampakoa!

@emackey @mramato please merge when ready.

mramato commented:
I apologize for the delay here @bampakoa.  Doing a final review now.

mramato commented on Apps/Sandcastle/gallery/Canvas Coordinates.html:
```diff
@@ -0,0 +1,58 @@
+    var result = viewer.scene.cartesianToCanvasCoordinates(position, tmp);
+    if(result){
+        logo.style.display = 'block';
+        logo.style.bottom = tmp.y + 'px';
```
This need to be `top` not `bottom`.

mramato commented on Apps/Sandcastle/gallery/Canvas Coordinates.html:
```diff
@@ -0,0 +1,58 @@
+logo.style.position = 'absolute';
+
+var handler = new Cesium.ScreenSpaceEventHandler(viewer.scene.canvas);
+handler.setInputAction(function(movement) {
```
Using `ScreenSpaceEventHandler` and mouse movement isn't this best approach here either (because the position isn't updated for non-mouse movement, like camera flights).  You would normally put code like this in the `scene.preRender` event.

mramato commented on Apps/Sandcastle/gallery/Canvas Coordinates.html:
```diff
@@ -0,0 +1,58 @@
+var viewer = new Cesium.Viewer('cesiumContainer');
+
+var logo = document.getElementById('logo');
+viewer.container.appendChild(logo);
```
The logo element is already in the dom, this line isn't needed.

mramato commented on Apps/Sandcastle/gallery/Canvas Coordinates.html:
```diff
@@ -0,0 +1,58 @@
+
+var logo = document.getElementById('logo');
+viewer.container.appendChild(logo);
+logo.style.position = 'absolute';
```
This should go in the HTML

mramato commented on Apps/Sandcastle/gallery/Canvas Coordinates.html:
```diff
@@ -0,0 +1,58 @@
+    var tmp = new Cesium.Cartesian2();
+    var result = viewer.scene.cartesianToCanvasCoordinates(position, tmp);
+    if(result){
+        logo.style.display = 'block';
```
This isn't needed, you can just have it always showing, it will end up positioned offscreen.

mramato commented:
`Canvas Coordinates.JPG` should be `Canvas Coordinates.jpg`

mramato commented:
@bampakoa since I dragged my feet on this and your core code was ready to merge, I went ahead and addressed the Sandcastle comments I made and also renamed it to `HTML Overlays`, since I feel that's more discoverable.

Thanks again for this PR and sorry for the delay!

bampakoa commented:
@mramato I think I need to practice more on sandcastle demos :smiley: I am glad I helped guys! Thanks very much!

mramato commented:
And I need to practice staying up to date on pull requests 😄 Thanks again!

pjcozzi commented:
Thanks again for the contribution, @bampakoa!  It is much appreciated.

@mramato could you please update CHANGES.md in master?
